File: fidimag/atomistic/demag_full.py
import fidimag.extensions.clib as clib
import numpy as np
from .energy import Energy


class DemagFull(Energy):
    """
    Calculation of the demag field using a brute-force approach,
    summing the dipolar contributions of the whole system for
    every lattice point.
    Since we can obtain the field directly, the energy is calculated
    in this process, thus we inherit from the Energy class
    as in the Exchange field, to calculate the total energy summing up
    the energy density.
    """

    def __init__(self, name='DemagFull'):
        self.name = name
        self.jac = True

    def setup(self, mesh, spin, mu_s, mu_s_inv):
        super(DemagFull, self).setup(mesh, spin, mu_s, mu_s_inv)

        unit_length = mesh.unit_length
        self.mu_s_scale = np.zeros(mesh.n, dtype=np.float64)

        # note that the 1e-7 comes from \frac{\mu_0}{4\pi}
        self.scale = 1e-7 / (unit_length ** 3)

        # could be wrong, needs carefully tests!!!
        self.mu_s_scale = self.mu_s * self.scale

    def compute_field(self, t=0, spin=None):
        if spin is not None:
            m = spin
        else:
            m = self.spin

        clib.compute_demag_full(m, self.field,
                                self.energy,
                                self.mesh.coordinates,
                                self.mu_s,
                                self.mu_s_scale,
                                self.n
                                )

        return self.field

    # def compute_energy(self):
    #     energy = self.demag.compute_energy(
    #         self.spin, self.mu_s_scale, self.field)
    #     return energy / self.scale

File: fidimag/atomistic/lib/demag_full.c
#include "clib.h"
#include "math.h"
#include "stdlib.h"

void demag_full(double *spin, double *field, double *coords,
                double *energy, int n) {

    /* Full calculation of Demag */

	#pragma omp parallel
	for (int i = 0; i < n; i++) {
        double* rij = malloc(3 * sizeof(double));
        double* rij_n = malloc(3 * sizeof(double));
        double mrij = 0;
        double rij_mag;

        for (int j = 0; j < n; j++) {

            if(j != i){
                
                for(int k = 0; k < 3; k++) {
                    rij[k] = coords[3 * j + k] - coords[3 * i + k];
                }

                rij_mag = sqrt(rij[0] * rij[0] + 
                               rij[1] * rij[1] + 
                               rij[2] * rij[2]);

                for(int k = 0; k < 3; k++) rij_n[k] = rij[k] / rij_mag;

                mrij = spin[3 * j] * rij[0] + spin[3 * j + 1] * rij[1]
                       + spin[3 * j + 2] * rij[2] ;

                for(int k = 0; k < 3; k++){
                    field[3 * i + k] += (3 * rij_n[k] * mrij - spin[3 * j + k])
                                         / (rij_mag * rij_mag * rij_mag) ;
                }
            }
        }
        free(rij);
        free(rij_n);
    }
}
